### ISSUE TITLE
Function declaration at block scope is bound in surrounding scope

### ISSUE THREAD

**Issue by mcmlevi** [llvm/llvm-project#64528]
So recently I ran into a bug on [MSVC](https://developercommunity.visualstudio.com/t/Forward-declarations-inside-of-a-functio/10432628) , during testing and research into this issue me and some other folks ran into a issue that's somewhat similar. 

If you take the following code sample this shouldn't compile as specified in the following draft[https://eel.is/c++draft/dcl.dcl#dcl.meaning.general-3.5] 

> If the declaration inhabits a block scope S and declares a function ([[dcl.fct]](https://eel.is/c++draft/dcl.dcl#dcl.fct)) or uses the extern specifier, the declaration shall not be attached to a named module ([[module.unit]](https://eel.is/c++draft/module.unit)); its target scope is the innermost enclosing namespace scope, but the name is bound in S[.](https://eel.is/c++draft/dcl.dcl#dcl.meaning.general-3.5.sentence-1)

```cpp
namespace Foo
{   
	void SomeOuterFunc()
	{
		void someFunc(int);
		someFunc(0);
	}
}

void Foo::someFunc(int)
{
}


int main() {
    Foo::SomeOuterFunc();
}
```

So with that, this shouldn't compile on Clang and should produce an error similar to GCC
```
<source>:10:6: warning: 'void Foo::someFunc(int)' has not been declared within 'Foo'
   10 | void Foo::someFunc(int)
      |      ^~~
<source>:5:22: note: only here as a 'friend'
    5 |                 void someFunc(int);
      |       
 ```` 

or MSVC 
```
1>main.cpp(17,11): error C2039: 'someFunc': is not a member of 'Foo'
1>main.cpp(8): message : see declaration of 'Foo'
```

**Comment by Eisenwave** [comment]
It's worth noting that clang partially fails to reject [[dcl.meaning.general] Example 3](https://eel.is/c++draft/dcl.meaning.general#example-3):
```cpp
namespace X {
  void p() {
    q();                        // error: q not yet declared
    extern void q();            // q is a member of namespace X
    extern void r();            // r is a member of namespace X
  }

  void middle() {
    q();                        // error: q not found
  }

  void q() { /* ... */ }        // definition of X​::​q
}

void q() { /* ... */ }          // some other, unrelated q
void X::r() { /* ... */ }       // error: r cannot be declared by qualified-id
```
The last line should produce an error, but it does not. See https://godbolt.org/z/enzYWc8qz

**Comment by Eisenwave** [comment]
This is a regression; clang 3.3 used to reject this with the message:
```
<source>:16:9: error: out-of-line definition of 'r' does not match any declaration in namespace 'X'
void X::r() { /* ... */ }       // error: r cannot be declared by qualified-id
        ^
```
See https://godbolt.org/z/zera4v4sv

**Comment by shafik** [comment]
It looks like we are missing the diagnostic for this case b/c in `Sema::ActOnFunctionDeclarator` this line results in `false`:

https://github.com/llvm/llvm-project/blob/8f4a0b701f32deb66da18fec63a90bf747a6b41f/clang/lib/Sema/SemaDecl.cpp#L10590-L10591

**Comment by shafik** [comment]
I am not sure if `D.getCXXScopeSpec().isSet()` should be `true` or we need to add an additional check to catch this case.

CC @zygoloid
